File: DistanceKBinaryTree.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;

	TreeNode(int x) {
		val = x;
	}
}

public class DistanceKBinaryTree {

	public static List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
		Map<TreeNode, TreeNode> parent = new HashMap<>();
		Queue<TreeNode> queue = new LinkedList<>();

		dfs(root, null, parent);

		Set<TreeNode> visited = new HashSet<>();

		queue.add(target);
		visited.add(target);

		int level = 0;

		// Typical BFS
		while (!queue.isEmpty()) {
			if (level == k)
				break;
			int size = queue.size();
			for (int i = 0; i < size; i++) {
				TreeNode node = queue.poll();

				if (node.left != null && !visited.contains(node.left)) {
					visited.add(node.left);
					queue.offer(node.left);
				}

				if (node.right != null && !visited.contains(node.right)) {
					visited.add(node.right);
					queue.offer(node.right);
				}

				TreeNode par = parent.get(node);

				if (par != null && !visited.contains(par)) {
					visited.add(par);
					queue.offer(par);
				}
			}

			level++;
		}

		List<Integer> res = new ArrayList<>();
		for (TreeNode i : queue) {
			res.add(i.val);
		}

		return res;
	}

	private static void dfs(TreeNode node, TreeNode par, Map<TreeNode, TreeNode> parent) {
		if (node == null)
			return;

		parent.put(node, par);
		dfs(node.left, node, parent);
		dfs(node.right, node, parent);
	}

	public static void main(String[] args) {

	}
}

File: distance_k_binary_tree.py
# for reference : https://bit.ly/3r9ZNfJ
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def buildParentMap(self, node, parent, parent_map):
        if not node:
            return
        parent_map[node] = parent
        self.buildParentMap(node.left, node, parent_map)
        self.buildParentMap(node.right, node, parent_map)

    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        parent_map = {}
        self.buildParentMap(root, None, parent_map)

        visited = set()
        res = []

        def dfs(node, dist):
            if not node:
                return
            if node in visited:
                return
            visited.add(node)
            if dist == k:
                res.append(node.val)
            elif dist < k:
                dfs(node.left, dist+1)
                dfs(node.right, dist+1)
                dfs(parent_map[node], dist+1)

        dfs(target, 0)
        return res
